stubs/ldap3/ldap3/core/connection.pyi

```diff
@@ -0,0 +1,169 @@
+from types import TracebackType
+from typing import Any, Type
+from typing_extensions import Literal
+
+from .server import Server
+
+SASL_AVAILABLE_MECHANISMS: Any
+CLIENT_STRATEGIES: Any
+
+class Connection:
+    connection_lock: Any
+    last_error: str
+    strategy_type: Any
+    user: Any
+    password: Any
+    authentication: Any
+    version: Any
+    auto_referrals: Any
+    request: Any
+    response: Any | None
+    result: Any
+    bound: bool
+    listening: bool
+    closed: bool
+    auto_bind: Any
+    sasl_mechanism: Any
+    sasl_credentials: Any
+    socket: Any
+    tls_started: bool
+    sasl_in_progress: bool
+    read_only: Any
+    lazy: Any
+    pool_name: Any
+    pool_size: Any
+    cred_store: Any
+    pool_lifetime: Any
+    pool_keepalive: Any
+    starting_tls: bool
+    check_names: Any
+    raise_exceptions: Any
+    auto_range: Any
+    extend: Any
+    fast_decoder: Any
+    receive_timeout: Any
+    empty_attributes: Any
+    use_referral_cache: Any
+    auto_escape: Any
+    auto_encode: Any
+    source_address: Any
+    source_port_list: Any
+    server_pool: Any | None
+    server: Any
+    strategy: Any
+    send: Any
+    open: Any
+    get_response: Any
+    post_send_single_response: Any
+    post_send_search: Any
+    def __init__(
+        self,
+        server: Server | str,
+        user: str | None = ...,
+        password: str | None = ...,
+        auto_bind: Literal["DEFAULT", "NONE", "NO_TLS", "TLS_BEFORE_BIND", "TLS_AFTER_BIND"] = ...,
+        version: int = ...,
+        authentication: Literal["ANONYMOUS", "SIMPLE", "SASL", "NTLM"] | None = ...,
+        client_strategy: Literal[
+            "SYNC", "SAFE_SYNC", "ASYNC", "LDIF", "RESTARTABLE", "REUSABLE", "MOCK_SYNC", "MOCK_ASYNC", "ASYNC_STREAM"
+        ] = ...,
+        auto_referrals: bool = ...,
+        auto_range: bool = ...,
+        sasl_mechanism: str | None = ...,
+        sasl_credentials: Any | None = ...,
+        check_names: bool = ...,
+        collect_usage: bool = ...,
+        read_only: bool = ...,
+        lazy: bool = ...,
+        raise_exceptions: bool = ...,
+        pool_name: str | None = ...,
+        pool_size: str | None = ...,
+        pool_lifetime: int | None = ...,
+        cred_store: Any | None = ...,
+        fast_decoder: bool = ...,
+        receive_timeout: Any | None = ...,
+        return_empty_attributes: bool = ...,
+        use_referral_cache: bool = ...,
+        auto_escape: bool = ...,
+        auto_encode: bool = ...,
+        pool_keepalive: Any | None = ...,
+        source_address: str | None = ...,
+        source_port: int | None = ...,
+        source_port_list: Any | None = ...,
+    ) -> None: ...
+    def repr_with_sensitive_data_stripped(self): ...
+    @property
+    def stream(self): ...
+    @stream.setter
+    def stream(self, value) -> None: ...
+    @property
+    def usage(self): ...
+    def __enter__(self) -> Connection: ...
```

Comment:
Shouldn't this use `typeshed.Self`?

Reply:
`_typeshed.Self` isn't yet usable in third-party stubs as mypy 0.910 doesn't include it. But I've declared a custom typevar for now.

Reply:
There is a lot of code already using it:

```
(env) akuli@akuli-desktop:~/typeshed/stubs$ git grep 'from _typeshed import Self' | wc -l
13
```

Reply:
Yeah, we should probably not have done that. It's not the end of the world, though, as mypy just assumed `Any` (and we shouldn't change it, since hopefully mypy gets a new release soon).

The last mypy release is so long ago that it's hard to remember what it contained. :(